faunadb-scala/src/test/scala/faunadb/ClientSpec.scala

```diff
@@ -1616,9 +1627,371 @@ class ClientSpec
+      }
+
+      override def onError(t: Throwable): Unit = capturedEventsP.failure(t)
+
+      override def onComplete(): Unit = capturedEventsP.failure(new IllegalStateException("not expecting the stream to complete"))
+    }
+
+    // subscribe to publisher
+    publisherValue.subscribe(valueSubscriber)
+
+    // push 3 updates
+    client.query(Update(docRef, Obj("data" -> Obj("testField" -> "testValue1")))).futureValue
+    client.query(Update(docRef, Obj("data" -> Obj("testField" -> "testValue2")))).futureValue
+    client.query(Update(docRef, Obj("data" -> Obj("testField" -> "testValue3")))).futureValue
+
+    // blocking
+    capturedEventsP.future.futureValue
+  }
+
+  it should "stream handles authorization lost during stream evaluation." in {
+    // create collection
+    val collectionName = "stream-things"
+    adminClient.query(CreateCollection(Obj("name" -> collectionName))).futureValue
+
+    // create doc
+    val createdDoc = adminClient.query(Create(Collection(collectionName), Obj("credentials" -> Obj("password" -> "abcdefg")))).futureValue
+    val docRef = createdDoc(RefField)
+
+    // create new key + client
+    val newKey = adminClient.query(CreateKey(Obj("role" -> "server-readonly"))).futureValue
+    val streamingClient = client.sessionClient(newKey(SecretField).get)
+
+    val publisherValue = streamingClient.stream(docRef).futureValue
+
+    val subscriberDone = Promise[Unit]
+
+    val valueSubscriber = new Flow.Subscriber[Value] {
+      var subscription: Flow.Subscription = null
+      val captured = new ConcurrentLinkedQueue[Value]
+      override def onSubscribe(s: Flow.Subscription): Unit = {
+        subscription = s
+        subscription.request(1)
+      }
+      override def onNext(v: Value): Unit = {
+        if (captured.isEmpty) {
+          // update doc on `start` event
+          adminClient.query(Update(docRef, Obj("data" -> Obj("testField" -> "afterStart")))).futureValue
+
+          // delete key
+          adminClient.query(Delete(newKey(RefField))).futureValue
+
+          // push an update to force auth revalidation.
+          adminClient.query(Update(docRef, Obj("data" -> Obj("testField" -> "afterKeyDelete")))).futureValue
+        }
+        captured.add(v)          // capture element
+        subscription.request(1)  // ask for more elements
+      }
+      override def onError(t: Throwable): Unit = subscriberDone.failure(t)
+      override def onComplete(): Unit = subscriberDone.success(())
+    }
+
+    // subscribe to publisher
+    publisherValue.subscribe(valueSubscriber)
+
+    // blocking - we expect the Promise to fail because the stream's key has been deleted
+    val subscriberError = subscriberDone.future.failed.futureValue
+    subscriberError shouldBe a[PermissionDeniedException]
+    subscriberError.getMessage should include("Authorization lost during stream evaluation.")
+  }
+
+  it should "stream buffers events if the producer is faster than the consumer" in {
```

Comment:
It seems that this one should go into the load test suite as well... It would be ok if the test takes longer there, so we could check the buffer limit (256).

Reply:
same as above

Reply:
https://github.com/fauna/faunadb-jvm/pull/272/commits/68d724693fe17df04780928a4c5f809fe34af826